File: 560-subarray-sum-equals-k/560-subarray-sum-equals-k.ts
function subarraySum(nums: number[], K: number): number {
  const N = nums.length,
        dp = new Array(N + 1).fill(0);
  let ans = 0;
  
  for(let i = 0; i < N; i++)
    dp[i + 1] = dp[i] + nums[i];

  for(let i = 1; i <= N; i++)
    for(let j = i - 1; j >= 0; j--) {
      const diff = dp[i] - dp[j];
      if(diff == K)
        ans += 1;
      
      // if(diff > K)
      //   break;
    }
      
  return ans;
};

File: 560-subarray-sum-equals-k/560-subarray-sum-equals-k.java
class Solution {
    
    public int subarraySum(int[] nums, int K) {
        int prefix = 0, ans = 0;
        Map<Integer, Integer> sumOcc = new HashMap<>();
        
        sumOcc.put(0, 1);

        for (int num : nums) {
            prefix += num;
            ans += sumOcc.getOrDefault(prefix - K, 0);
            sumOcc.put(prefix, 1 + sumOcc.getOrDefault(prefix, 0));
        }
        
        return ans;
    }
}